### ISSUE TITLE
Convert Charmm .psf file to Gromacs .top file

### ISSUE THREAD

**Issue by MamtaMohan** [ParmEd/ParmEd#750]
Hello,

Can ParmEd convert .psf file to .top file.

Mamta

**Comment by swails** [comment]
Yes it can.  Do it something like this:

```python
import parmed as pmd

psf = pmd.load_file('your_file.psf')
psf.coordinates = pmd.load_file('your_file.crd').coordinates
params = pmd.charmm.CharmmParameterSet('file1.prm', 'file1.rtf', 'file1.str', ...)
psf.load_parameters(params)
psf.save('your_file.top') # .top is recognized as a GROMACS topology file
psf.save('your_file.gro')
```

You just have to make sure that you load all of the parameter, topology, and stream files necessary to assign parameters to your PSF file when you define your `CharmmParameterSet`.

Of course, change the file names to those relevant for your use case.

**Comment by MamtaMohan** [comment]
Dear Jason,

Is it .pdb file in psf.coordinates or file extention is .crd?

I tried .pdb file and I am getting this error:
In [1]: import parmed as pmd

In [2]: psf = pmd.load('1
1aki_autopsf.pdb  1aki_autopsf.psf  

In [2]: psf = pmd.load('1aki_autopsf.psf')
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-2-723d116c3ab6> in <module>()
----> 1 psf = pmd.load('1aki_autopsf.psf')

AttributeError: 'module' object has no attribute 'load'

In [3]: psf = pmd.load_file('1aki_autopsf.psf')

In [4]: psf.coordinates = pmd.load_file('1aki_autopsf.pdb').coordinates

In [5]: params = pmd.charmm.CharmmParameterSet('/usr/local/lib/vmd/plugins/noarch/tcl/readcharmmtop1.1/top_all27_prot_lipid_na.inp')

In [6]: psf.load_parameters(params)
---------------------------------------------------------------------------
ParameterError                            Traceback (most recent call last)
<ipython-input-6-44af4676212a> in <module>()
----> 1 psf.load_parameters(params)

/home/mamta/anaconda2/lib/python2.7/site-packages/parmed/charmm/psf.pyc in load_parameters(self, parmset, copy_parameters)
    539                 bond.type = parmset.bond_types[key]
    540             except KeyError:
--> 541                 raise ParameterError('Missing bond type for %r' % bond)
    542             bond.type.used = False
    543         # Build the bond_types list

ParameterError: Missing bond type for <Bond <Atom N [0]; In LYS 0>--<Atom CA [4]; In LYS 0>; type=None>

In [7]:

**Comment by MamtaMohan** [comment]
Dear Jason,

I seems to get an error during conversion.

First error I posted up there. But it was due to .pdb I am trying to read but function expects .crd (charmm format file).

Realizing this I converted .pdb file to .crd file using Bio3D (an R package) to tried the command again.

Still I got this error:
In [10]: psf.coordinates = pmd.load_file('1aki.crd').coordinates---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-10-87be71bf69c7> in <module>()
----> 1 psf.coordinates = pmd.load_file('1aki.crd').coordinates

/home/mamta/anaconda2/lib/python2.7/site-packages/parmed/structure.pyc in coordinates(self, value)
   1676             value = list(value)
   1677             coords = np.array(value, dtype=np.float64, copy=False, subok=True)
-> 1678             coords = coords.reshape((-1, len(self.atoms), 3))
   1679             if len(coords) > 0:
   1680                 for a, xyz in zip(self.atoms, coords[0]):

ValueError: total size of new array must be unchanged

Please let me know what I am missing.

Mamta

**Comment by swails** [comment]
You don't need to change to a crd file.  `pmd.load_file` will automatically detect the file format.  So it will work equally well with a CHARMM crd file or a PDB file.

The error message you received indicates that your coordinate file has a different number of atoms than your PSF file.  Notice that `coords.reshape((-1, len(self.atoms), 3))` will reshape the coordinate array into a list of coordinate frames having X, Y, and Z components for each atom.  If the coordinate length is not a multiple of # of atoms * 3, this reshaping will fail.

Basically, your coordinate file and PSF file are not compatible, and you need to fix that.

**Comment by swails** [comment]
Also, the first error you got was not because of using a PDB instead of a coordinate file, it was because your parameter set did not include any bond parameters.  You instantiated the parameter set from just a single topology file.  Topology files in CHARMM do not have any parameters -- you *also* need to include the parameter file.

Also, you will probably need to include a stream file for the MTSL moiety.

**Comment by MamtaMohan** [comment]
Dear Jason,

Thank you for your response.

I will do that.

**Comment by MamtaMohan** [comment]
Dear Jason,

I am still having issues. I gave all the parameter files I used to build .psf in NAMD.
But I think I am not giving the Charmm files properly. Let me understand:

To convert .psf to .top - I have to give .prm files. Do I have to give .rtf and .str files as well.

My system in NAMD is build against files:
par_all36_carb.prm 
par_all36_cgenff.prm 
par_all36_lipid.prm 
par_all36_na.prm 
par_all36_prot.prm 
par_all27_prot_lipid_na.inp 
par_all22_prot_cmap.inp 
toppar_all36_spin_label.inp 
toppar_all36_spin_label_dummy.inp 

Currently I am using protein 1aki.pdb to get a test script. It does not have spin label.

I will appreciate your response.

Mamta

**Comment by swails** [comment]
What have you tried?

Your `CharmmParameterSet` should take every file above in the constructor:

```python
import parmed as pmd

...
params = pmd.charmm.CharmmParameterSet('/path/to/par_all36_carb.prm', '/path/to/par_all36_cgenff.prm', ...)
```

**Comment by MamtaMohan** [comment]
No. It is not the problem of taking file. It does.

I am getting error:

 psf.load_parameters(params)
---------------------------------------------------------------------------
ParameterError                            Traceback (most recent call last)
<ipython-input-9-44af4676212a> in <module>()
----> 1 psf.load_parameters(params)

/home/mamta/anaconda2/lib/python2.7/site-packages/parmed/charmm/psf.pyc in load_parameters(self, parmset, copy_parameters)
    525             except KeyError:
    526                 raise ParameterError('Could not find atom type for %s' %
--> 527                                      atom.type)
    528             atom.atom_type = atype
    529             # Change to string type to look up the rest of the parameters

ParameterError: Could not find atom type for HB

So I know I am missing something.

**Comment by MamtaMohan** [comment]
Following command worked.

params = pmd.charmm.CharmmParameterSet('/home/mamta/test/TopoTools_test/TopoTools/Toppar/par_all22_prot_cmap.inp', '/home/mamta/test/TopoTools_test/TopoTools/Toppar/par_all36_carb.prm', '/home/mamta/test/TopoTools_test/TopoTools/Toppar/par_all36_cgenff.prm', '/home/mamta/test/TopoTools_test/TopoTools/Toppar/par_all36_lipid.prm', '/home/mamta/test/TopoTools_test/TopoTools/Toppar/par_all36_na.prm', '/home/mamta/test/TopoTools_test/TopoTools/Toppar/par_all36_prot.prm')

**Comment by MamtaMohan** [comment]
I am sorry. I misspoke. Above command put out warnings like:
 params = pmd.charmm.CharmmParameterSet('/home/mamta/test/TopoTools_test/TopoTools/Toppar/par_all22_prot_cmap.inp', '/home/mamta/test/TopoTools_test/TopoTools/Toppar/par_all36_carb.prm', '/home/mamta/test/TopoTools_test/TopoTools/Toppar/par_all36_cgenff.prm', '/home/mamta/test/TopoTools_test/TopoTools/Toppar/par_all36_lipid.prm', '/home/mamta/test/TopoTools_test/TopoTools/Toppar/par_all36_na.prm', '/home/mamta/test/TopoTools_test/TopoTools/Toppar/par_all36_prot.prm')
/home/mamta/anaconda2/lib/python2.7/site-packages/parmed/charmm/parameters.py:719: ParameterWarning: Atom type CPM not present in AtomType list
  ParameterWarning)
/home/mamta/anaconda2/lib/python2.7/site-packages/parmed/charmm/parameters.py:442: ParameterWarning: Replacing bond ('CA', 'CPT'), <BondType; k=305.000, req=1.368> with <BondType; k=300.000, req=1.360>
  (key, self.bond_types[key], bond_type), ParameterWarning)
/home/mamta/anaconda2/lib/python2.7/site-packages/parmed/charmm/parameters.py:442: ParameterWarning: Replacing bond ('CPT', 'CPT'), <BondType; k=360.000, req=1.400> with <BondType; k=360.000, req=1.385>
  (key, self.bond_types[key], bond_type), ParameterWarning)

**Comment by hainm** [comment]
@MamtaMohan  you can check the syntax highlight on github. It's easy to spot the code

https://help.github.com/articles/creating-and-highlighting-code-blocks/

```python
/home/mamta/anaconda2/lib/python2.7/site-packages/parmed/charmm/parameters.py:442: 
ParameterWarning: Replacing bond ('CPT', 'CPT'), with
(key, self.bond_types[key], bond_type), ParameterWarning)
```

**Comment by MamtaMohan** [comment]
Hi Hainm,

I do not think it is a syntax issue.

I think I am missing something while giving files to load parameters.

**Comment by hainm** [comment]
:D sorry not to be clear. I meant when you pasted your error log to github issue.

**Comment by MamtaMohan** [comment]
I am sorry I missed again what you are saying.

**Comment by MamtaMohan** [comment]
Dear Hainm,


I would take you suggestion during development project.


I am new to python and lot of things.


I am trying to work my way through all this.


But I appreciate your suggestions.


Mamta

________________________________

**Comment by swails** [comment]
@MamtaMohan -- what @hainm was saying is that if you do not format the raw input and output you get as 'raw text' (or even better, Python code), then it is much harder for us to visualize it (since a lot of the characters get interpreted in Markdown as adding bold, or headers, or italics).  So always paste your raw input/output in blocks that look like this:

\`\`\`python
\`\`\`

For example, look at the difference between this:

# Put in a main block in the script now
if __name__ == '__main__':
    # This only executes inside the main block
    print('I am in the main block!')

and this:

```python
# Put in a main block in the script now
if __name__ == '__main__':
    # This only executes inside the main block
    print('I am in the main block!')
```

The second is clearly much easier to read than the second.  @hainm was not giving you advice specifically on how to fix your problem, just on how to report your problems in a way that is easier for us to understand.

(Also, I have gone back to some of your posts and added the \`\`\`python markers myself to make it easier to read).

**Comment by swails** [comment]
As for your problems -- whenever you see a `ParameterWarning` or a `MissingParameter` statement, then you have not loaded the parameter files that contain the necessary parameters.  If you can find the parameter file you loaded that has the parameters that ParmEd complains are missing, point that out to me and it's a bug that needs to be fixed.

But as long as your error says something about a missing parameter, my answer will be the same: you need to load all of the necessary parameter, topology, and stream files for your system.  Only you can know what those truly are.

**Comment by MamtaMohan** [comment]
Dear Jason,

Thank you.

Now I understand.
I will do that in future.
Please pardon me for mentioning that I am new to gromacs too.

Generating .top via gromacs is easy if it is clean system just protein or so. 
But my system contain spin label so I have to take ParmEd route to convert .PSF to .top file.

Now I would like to confirm to generate .top file besides .prm file I have to give .str and .rtf files as well to build parameters.

**Comment by swails** [comment]
The steps you need to take for ParmEd are the following:

1. Read in the CHARMM input files (PSF, CRD or PDB, and parameter files) so that the molecule is completely described with all parameters within ParmEd
1. Save this file in GROMACS format (both coordinates and topology)

ParmEd will automatically take care of assigning the appropriate parameters inside the GROMACS topology file based on the contents of the CHARMM parameter files.  If you don't provide all of the parameter files in the first step, though, you can't move on to the second.

Another possible route you can take is to use CHARMM-GUI, which will generate GROMACS input files directly for you.

**Comment by MamtaMohan** [comment]
I will take ParmEd route. Because Charmm-Gui deals with only small molecules. Usually ligand. If I remember correctly. I used Charmm-Gui to generate spin labels parameters for .psf as well gromacs.

But I will investigate it again in case I am missing something.

Truly appreciate your response.

Mamta

**Comment by swails** [comment]
> Because Charmm-Gui deals with only small molecules.

Definitely not true :).  CHARMM-GUI can handle anything that CHARMM itself can.  You may find it easier to use CHARMM-GUI than ParmEd (although possibly not, because you still have the problem of needing to get all of the required parameter and topology files).  Of course, CHARMM-GUI generates the PSF itself -- you can't give it a PSF you already generated with VMD.

**Comment by MamtaMohan** [comment]
Thank you Jason.

I did generated files using Charmm-gui for NAMD system with labels but to generate them for Gromacs I have to generate.top from .psf and I think ParmEd will help me there.

**Comment by swails** [comment]
CHARMM-GUI gives you a zip file with a gromacs folder that contains the files.  The conversion is done automatically for you.

**Comment by MamtaMohan** [comment]
Dear Jason,

Thank you for your response.

Charmm-gui does give a folder with all the .prm, .rtf and .str file in it.

I am still having issue with loading parameters and parsed to PDB file.

For protein 3lzm.pdb. All ideally I need par_all36_prot.prm and top_all36_prot.rtf. It does not have any label or ligand and it should go clean.

However I got error:

```
/home/mamta/anaconda2/lib/python2.7/site-packages/parmed/charmm/psf.pyc in load_parameters(self, parmset, copy_parameters)
    525             except KeyError:
    526                 raise ParameterError('Could not find atom type for %s' %
--> 527                                      atom.type)
    528             atom.atom_type = atype
    529             # Change to string type to look up the rest of the parameters

ParameterError: Could not find atom type for HB

```
So I brought in other .prm and .rtf files and finally toppar_water_ions.str file. Still the same error.

So at this point I decided to write to you. It does not make sense to me. Unless there is something with psfgen or charmm36 FF itself.


```
import parmed as pmd
psf = pmd.load_file('3lzm_autopsf.psf')
psf.coordinates = pmd.load_file('3lzm_autopsf.pdb').coordinates
params = pmd.charmm.CharmmParameterSet('../toppar/par_all36_prot.prm', '../toppar/top_all36_prot.rtf', '../toppar/par_all36_carb.prm', '../toppar/top_all36_carb.rtf', '../toppar/par_all36_cgenff.prm', '../toppar/top_all36_cgenff.rtf', '../toppar/par_all36_lipid.prm', '../toppar/top_all36_lipid.rtf', '../toppar/par_all36_na.prm', '../toppar/top_all36_na.rtf', '../toppar/toppar_water_ions.str')
psf.load_parameters(params)

/home/mamta/anaconda2/lib/python2.7/site-packages/parmed/charmm/psf.pyc in load_parameters(self, parmset, copy_parameters)
    525             except KeyError:
    526                 raise ParameterError('Could not find atom type for %s' %
--> 527                                      atom.type)
    528             atom.atom_type = atype
    529             # Change to string type to look up the rest of the parameters

ParameterError: Could not find atom type for HB

```

Please guide me what I am doing wrong.

Mamta

**Comment by swails** [comment]
> Charmm-gui does give a folder with all the .prm, .rtf and .str file in it.

That's not what I meant.  What I meant is that CHARMM should give a folder with a GROMACS topology and ITP files.  As much struggle as this seems to be causing, I think using CHARMM-GUI to create GROMACS files directly is going to be your easiest route.

> Please guide me what I am doing wrong.

I have no idea, and it's not entirely clear to me exactly what you are doing (are you using CHARMM-GUI? psfgen from VMD? CHARMM directly? what input are you providing to the programs?  how did you define the nonstandard residues? etc.).  The atom type `HB` is defined in one of the RTF files used by psfgen to create the PSF, but you are not using that RTF (or corresponding PRM) file in your ParmEd script.

I can't tell you *what* file that is because I have no way of knowing.

**Comment by MamtaMohan** [comment]
I understand. 

I appreciate all your help and response.

Thank you.
Mamta

**Comment by MamtaMohan** [comment]
Dear Jason,

Please pardon me.

I do not mean to end response abruptly but I thing I will explain to you later in mail what is it I am trying to do and what is going on.

I do truly appreciate all your help.

Thank you again.
Mamta

**Comment by MamtaMohan** [comment]
Dear Josh,

I would like to explain what was going on.

Issue #1: Charmm-gui engine and Gromacs input.
I was getting this error when I used .psf file generated via VMD psfgen program. A set of simulations has already been done used this file via NAMD. So to keep the things in sync I have to use this .psf file. To prepare this file a while back I had used Charmm-gui only to attach spin labels and get spin label parameters. For this I had used only Input generator. After Input generator I got a tar file from charmm-gui and from this point onwards I prepared system by myself. Since I never did any solvation, ionization and dynamics via Charmm-gui, the files I got never had any Gromacs input files, the option for which comes towards the end when we want to do simulation via Charmm-gui.
So it was very confusing for me when you mentioned couple of times about Charmm-gui giving me gromacs files. Today I went to Charmm-gui, picked a protein and went all the way to simulation and option for Gromacs input was given along with others options. That explained to me, my confusion and your comments. I can not use Charmm-gui to generate .top file because it does not satisfy the condition to which system is subjected. So I will politely decline your suggestion to use Charmm-gui to generate .top file via that. Instead I would be inclined to use ParmEd with the .psf file I have. 

Issue# 2: ParameterError: Could not find atom type for HB
ParmEd gave me error with Charmm36 files with .psf I was using. I realized yesterday after reading some VMD posts that there was some change when Charmm27 was brought forward to Charmm36(At the end you will see the links of those mails). When I checked VMD it brings in:  top_all27_prot_lipid_na.inp file. Since I was giving Charmm36 files when I was using ParmEd I got the error: Could not find atom type for HB. This explains the error. Since at the time I wrote to you I was not aware of the conversion issue I keep thinking that I am giving wrong input. Being new to Python, Gromacs and lot of other things further clouded the issue to me.
Thread: http://www.ks.uiuc.edu/Research/namd/mailing_list/namd-l.2012-2013/3273.html
Thread: http://www.ks.uiuc.edu/Research/namd/mailing_list/namd-l.2012-2013/2998.html
Thread: http://www.ks.uiuc.edu/Research/vmd/mailing_list/vmd-l/23118.html

This issue is closed but I feel compelled to explain you what was going on and want to thank you generously for your patience and help.

I look forward to work with you in future.

Mamta
